File: src/ui/markdown.ts
/**
 * Minimal, dependency-free Markdown → HTML for the dashboard preview.
 *
 * The whole document is HTML-escaped up front, so no path through the parser
 * can emit source-supplied markup: a `<script>` in a note renders as text.
 * Block markers survive escaping unchanged except `>`, which becomes `&gt;`.
 * Unknown input is never an error — worst case it comes back as paragraphs.
 */
import { takeCode, takeParagraph, takeQuote } from "./markdown-blocks.js";
import { frontmatterBlock, frontmatterRows } from "./markdown-frontmatter.js";
import { escapeHtml, renderInline } from "./markdown-inline.js";
import { LIST_RE, renderList } from "./markdown-list.js";
import { isTableSeparator, takeTable } from "./markdown-table.js";

export { renderFrontmatter } from "./markdown-frontmatter.js";

const FENCE = /^\s*```\s*([\w+#.-]*)/;
const HEADING = /^(#{1,6})\s+(.*)$/;
const RULE = /^(-{3,}|\*{3,}|_{3,})\s*$/;

/**
 * Frontmatter is the caller's business (see journal-scan); it never renders.
 *
 * Only a block that actually PARSES as key/value frontmatter is consumed —
 * the same test renderFrontmatter uses to decide whether there is a table to
 * show. Stripping on the delimiters alone deleted everything up to the second
 * `---` in any text that merely opened with a horizontal rule, and rendered
 * no table in its place, so a model reply beginning with `---` silently lost
 * its first section on the way through /api/render.
 */
export function stripFrontmatter(src: string): string {
	const block = frontmatterBlock(src);
	if (block === null || frontmatterRows(block).length === 0) return src;
	const end = src.indexOf("\n---", src.indexOf("\n"));
	const nl = src.indexOf("\n", end + 1);
	return nl === -1 ? "" : src.slice(nl + 1);
}

/** True for any line that must interrupt a running paragraph. */
function startsBlock(line: string): boolean {
	return (
		FENCE.test(line) ||
		HEADING.test(line) ||
		RULE.test(line.trim()) ||
		line.startsWith("&gt;") ||
		LIST_RE.test(line)
	);
}

function takeList(lines: string[], start: number): [string, number] {
	let i = start;
	while (i < lines.length && LIST_RE.test(lines[i] ?? "")) i++;
	return [renderList(lines.slice(start, i)), i];
}

export function renderMarkdown(src: string): string {
	if (typeof src !== "string" || src === "") return "";
	const body = stripFrontmatter(src).replace(/\r\n?/g, "\n");
	// Escaping never adds or removes a newline, so `raw` and `lines` are the
	// same document line-for-line: index i means the same line in both. Fenced
	// code needs the unescaped text (the highlighter escapes as it wraps).
	const raw = body.split("\n");
	const lines = escapeHtml(body).split("\n");
	const out: string[] = [];
	let i = 0;
	while (i < lines.length) {
		const line = lines[i] ?? "";
		if (line.trim() === "") {
			i++;
			continue;
		}
		const fence = FENCE.exec(line);
		const heading = HEADING.exec(line);
		if (fence) {
			const [html, next] = takeCode(lines, i, fence[1] ?? "", raw);
			out.push(html);
			i = next;
		} else if (heading) {
			const n = (heading[1] ?? "#").length;
			out.push(`<h${n}>${renderInline((heading[2] ?? "").trim())}</h${n}>`);
			i++;
		} else if (RULE.test(line.trim())) {
			out.push("<hr>");
			i++;
		} else if (line.startsWith("&gt;")) {
			const [html, next] = takeQuote(lines, i);
			out.push(html);
			i = next;
		} else if (line.includes("|") && isTableSeparator(lines[i + 1] ?? "")) {
			const [html, next] = takeTable(lines, i);
			out.push(html);
			i = next;
		} else if (LIST_RE.test(line)) {
			const [html, next] = takeList(lines, i);
			out.push(html);
			i = next;
		} else {
			const [html, next] = takeParagraph(lines, i, startsBlock);
			out.push(html);
			i = next > i ? next : i + 1;
		}
	}
	return out.join("\n");
}

File: crates/flusk-core/src/render/markdown.rs
//! Minimal, dependency-free Markdown → HTML — the render_markdown entry.
//!
//! The whole document is HTML-escaped up front, so no path through the
//! parser can emit source-supplied markup: a `<script>` in a note renders as
//! text. Block markers survive escaping unchanged except `>`, which becomes
//! `&gt;`. Unknown input is never an error — worst case it comes back as
//! paragraphs.

use super::blocks::{take_code, take_paragraph, take_quote};
use super::escape::escape_html;
use super::frontmatter::strip_frontmatter;
use super::inline::render_inline;
use super::js::{is_js_ws, js_dot, js_trim};
use super::list::{parse_item, render_list};
use super::table::{is_table_separator, take_table};

/// `^\s*```\s*([\w+#.-]*)` → Some(info word) when the line opens a fence.
fn fence_lang(line: &str) -> Option<String> {
    let cs: Vec<char> = line.chars().collect();
    let mut i = 0;
    while i < cs.len() && is_js_ws(cs[i]) {
        i += 1;
    }
    if cs.len() < i + 3 || cs[i] != '`' || cs[i + 1] != '`' || cs[i + 2] != '`' {
        return None;
    }
    i += 3;
    while i < cs.len() && is_js_ws(cs[i]) {
        i += 1;
    }
    let start = i;
    while i < cs.len() && (cs[i].is_ascii_alphanumeric() || matches!(cs[i], '_' | '+' | '#' | '.' | '-')) {
        i += 1;
    }
    Some(cs[start..i].iter().collect())
}

/// `^(#{1,6})\s+(.*)$` → (level, text after the whitespace run).
fn heading(line: &str) -> Option<(usize, String)> {
    let cs: Vec<char> = line.chars().collect();
    let mut h = 0;
    while h < cs.len() && cs[h] == '#' {
        h += 1;
    }
    if h == 0 || h > 6 {
        return None;
    }
    let mut i = h;
    while i < cs.len() && is_js_ws(cs[i]) {
        i += 1;
    }
    if i == h {
        return None;
    }
    let rest: String = cs[i..].iter().collect();
    rest.chars().all(js_dot).then_some((h, rest))
}

/// `^(-{3,}|\*{3,}|_{3,})\s*$` over the TRIMMED line.
fn is_rule(trimmed: &str) -> bool {
    let cs: Vec<char> = trimmed.chars().collect();
    cs.len() >= 3 && matches!(cs[0], '-' | '*' | '_') && cs.iter().all(|&c| c == cs[0])
}

/// True for any line that must interrupt a running paragraph.
fn starts_block(line: &str) -> bool {
    fence_lang(line).is_some()
        || heading(line).is_some()
        || is_rule(js_trim(line))
        || line.starts_with("&gt;")
        || parse_item(line).is_some()
}

fn take_list(lines: &[&str], start: usize) -> (String, usize) {
    let mut i = start;
    while i < lines.len() && parse_item(lines[i]).is_some() {
        i += 1;
    }
    (render_list(&lines[start..i]), i)
}

pub fn render_markdown(src: &str) -> String {
    if src.is_empty() {
        return String::new();
    }
    // Frontmatter is the caller's business; strip BEFORE newline
    // normalization, as the reference does. `\r\n?` → `\n` pairs greedily
    // left-to-right, then converts every remaining lone `\r`.
    let body = strip_frontmatter(src).replace("\r\n", "\n").replace('\r', "\n");
    // Escaping never adds or removes a newline, so `raw` and `lines` are the
    // same document line-for-line: index i means the same line in both.
    // Fenced code needs the unescaped text (the highlighter escapes itself).
    let raw: Vec<&str> = body.split('\n').collect();
    let escaped = escape_html(&body);
    let lines: Vec<&str> = escaped.split('\n').collect();
    let mut out: Vec<String> = Vec::new();
    let mut i = 0;
    while i < lines.len() {
        let line = lines[i];
        if js_trim(line).is_empty() {
            i += 1;
            continue;
        }
        if let Some(lang) = fence_lang(line) {
            let (html, next) = take_code(&lines, i, &lang, &raw);
            out.push(html);
            i = next;
        } else if let Some((n, text)) = heading(line) {
            out.push(format!("<h{n}>{}</h{n}>", render_inline(js_trim(&text))));
            i += 1;
        } else if is_rule(js_trim(line)) {
            out.push("<hr>".to_string());
            i += 1;
        } else if line.starts_with("&gt;") {
            let (html, next) = take_quote(&lines, i);
            out.push(html);
            i = next;
        } else if line.contains('|') && is_table_separator(lines.get(i + 1).copied().unwrap_or("")) {
            let (html, next) = take_table(&lines, i);
            out.push(html);
            i = next;
        } else if parse_item(line).is_some() {
            let (html, next) = take_list(&lines, i);
            out.push(html);
            i = next;
        } else {
            let (html, next) = take_paragraph(&lines, i, starts_block);
            out.push(html);
            i = if next > i { next } else { i + 1 };
        }
    }
    out.join("\n")
}
